### PR TITLE
Fix composition of href exclude paths to account for JSON encoding and differing site/home URLs

### LINKED ISSUE #1163: Speculative Loading invalid key for a URL pattern error - exclude_paths filter
## Bug Description

Trying to simply exclude a URL pattern and running into what appears to be a bug with the `plsr_speculation_rules_href_exclude_paths` filter.

Getting this error:

> Invalid key "0" for a URL pattern object found.

image

## Steps to reproduce

1. Fresh WordPress Install with Speculative Loading plugin running. 
2. Add a filter with the following exclusion and URL pattern. Using example from WordPress.org docs ([source](https://wordpress.org/plugins/speculation-rules/#how%20can%20i%20prevent%20certain%20urls%20from%20being%20prefetched%20and%20prerendered%3F)).

```php
add_filter(
    'plsr_speculation_rules_href_exclude_paths',
    function ( array $exclude_paths ): array {
        $exclude_paths[] = '/go/*';
        return $exclude_paths;
    }
);
```

3. Log out of the site and test in Chrome DevTools. Error should flag. 

image

## Additional Context

 - PHP Version: 8.2
 - OS: macOS Sonoma 14.4.1
 - Browser: Chrome Version 123.0.6312.124
 - Plugin Version: 1.2.1

If I'm doing something wrong, feel free to clarify and close the issue. Thanks!

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

When filtering `plsr_speculation_rules_href_exclude_paths` to add additional paths, make sure the resulting de-duplicated array has sequential keys starting from zero. If this is not the case, then `json_encode()` interprets the array as an object which causes a speculation rules failure in the browser.

Additionally, we found that the base exclude hrefs were getting prefixed with the site URL path and then prefixed again with the home URL path with the merge with the additional filtered exclude href exclude paths. On sites with different site URL and home URL (e.g. `/wp/` and `/blog/`) this can result in the base URLs getting double-prefixed in an erroneous prefix (e.g. `/blog/wp/`). This is fixed by avoiding the redundant prefixing, so that only the href exclude paths provided by the filter are prefixed with the home URL path. This PR also ensures that the nonce/`_wpnonce` action URLs (from #1143) are prefixed with the home URL path and not the site URL path.

Fixes #1163

## Relevant technical choices

The result of calling `array_unique()` is passed into `array_values()` to ensure that the keys are numerically sequential from zero. Tests are updated to ensure this is the case.

Additionally, the `$base_href_exclude_paths` are no longer provided as the initial value to pass into the `plsr_speculation_rules_href_exclude_paths` filter. The values in `$base_href_exclude_paths` get merged again with the result of the filter, meaning they will always be present regardless. It is misleading to pass them into the filter, as a developer may think they can remove a path by eliminating it from the array passed to the filter when they cannot. Eliminating this redundancy alone fixes the symptom of non-sequential array keys, but the changes in this PR go further to ensure there is no possibility of non-sequential array keys when the the base exclude paths are merged with the return value from the filter.

### CODE REVIEW

**Comment by westonruter** [comment]
Build for testing: [speculation-rules.zip](https://github.com/WordPress/performance/files/15033767/speculation-rules.zip)

**Review by tunetheweb — APPROVED** [review]
LGTM form a spec rules perspective but will leave others to comment on the PHP side of things.

**Review by adamsilverstein — APPROVED** [review]
Changes look good, thanks for the tests!

File: plugins/speculation-rules/helper.php [hunk 1]
```diff
@@ -57,18 +56,24 @@ function plsr_get_speculation_rules() {
...
+	// Ensure that:
+	// 1. There are no duplicates.
+	// 2. The base paths cannot be removed.
+	// 3. The array has sequential keys (i.e. array_is_list()).
+	$href_exclude_paths = array_values(
+		array_unique(
+			array_map(
+				array( $prefixer, 'prefix_path_pattern' ),
+				array_merge(
+					$base_href_exclude_paths,
```
Comment (joemcgill) [inline plugins/speculation-rules/helper.php]: Now that `$base_href_exclude_paths` are being run through `$prefixer::prefix_path_pattern` here, should we avoid doing so above?

**Review by joemcgill — APPROVED** [review]
One question, but looks good.

File: plugins/speculation-rules/helper.php (hunk 1 above)
Comment (westonruter) [reply plugins/speculation-rules/helper.php]: Oh! I think you identified another bug, specifically when the `site_url` and `home_url` are different. I've added a test case to show this: 74e52a21 & d77b46c1. The problem is that the base exclude paths for wp-admin and wp-login are initially prefixed with the site URL, but then later after merging with the filter-supplied exclude paths, it is getting prefixed _again_ and this time with the home URL. The result is the prefix from both the home URL _and_ the site URL is getting applied to the base URLs. I've proposed a fix for this in 01d9153aa7701d02198ac90c1102479345e7c449. Please check.

**Comment by westonruter** [comment]
Updated build for testing: [speculation-rules.zip](https://github.com/WordPress/performance/files/15043596/speculation-rules.zip)

**Review by joemcgill — APPROVED** [review]
This update looks good to me. Thanks for testing!

**Pull request merged**